crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1428,7 +1433,30 @@ impl<'db> Type<'db> {
                         .member(db, "__dict__");
                 }
 
-                let global_lookup = symbol(db, global_scope(db, *file), name);
+                // If the file that originally imported the module has also imported a submodule
+                // named [name], then the result is (usually) that submodule, even if the module
+                // also defines a (non-module) symbol with that name.
+                //
+                // Note that technically, either the submodule or the non-module symbol could take
+                // priority, depending on the ordering of when the submodule is loaded relative to
+                // the parent module's `__init__.py` file being evaluated.  That said, we have
+                // chosen to always have the submodule take priority.  (This matches pyright's
+                // current behavior, and opposite of mypy's current behavior.)
+                if let Some(submodule_name) = ModuleName::new(name) {
+                    let importing_file = module_ref.importing_file(db);
+                    let importing_index = semantic_index(db, importing_file);
+                    let mut full_submodule_name = module_ref.module(db).name().clone();
+                    full_submodule_name.extend(&submodule_name);
+                    if importing_index.imports_module(&full_submodule_name) {
+                        if let Some(submodule) = resolve_module(db, &full_submodule_name) {
+                            let submodule_ty = Type::module_literal(db, importing_file, submodule);
+                            return Symbol::Type(submodule_ty, Boundness::Bound);
+                        }
+                    }
```

Comment:
This is more of a fine-tuning comment: 
It could make sense to define an `imported_modules(db, file)` query that returns the set with all known imported modules (it would just be a thin wrapper around `semantic_index(db, file).imported_modules()`. 

The advantage of having this intermediate query is that currenlty, `Type::member` has to re-execute whenever the AST of `importing-file` changes. Salsa might to truncate any dependent queries because it can determine that the type returned by `Type::member` is still the same and, therefore, dependent queries don't have to re-run. However, salsa could determine this sooner if `Type::member` only dependent on a `imported_modules` query. It would then only have to re-run the `imported_modules` query and, if the imports are unchanged, can't stop immediately.

Reply:
Done, thanks for the suggestion!  I wrapped the `imported_modules` set in an `Arc`, since it seems like the set now needs to be able to outlive the containing `SemanticIndex`, is that right?